### PR TITLE
Amazon VPC CNI: Upgrade to v1.0 and Allow Custom Images

### PR DESCRIPTION
**Pull request by ripta**
We use the Amazon's VPC CNI plugin (`amazonvpc` in the cluster spec), which is currently bundled at version 0.1.1. The intent for this PR are:

1. To allow the docker image to be overridden in the cluster spec, which would allow easier testing to network upgrades or custom image builds. The spec snippet looks like:

```yaml
spec:
  networking:
    amazonvpc:
      imageName: repo.example.com/amazon-k8s-cni:0.1.1
```

2. To upgrade the default image version from 0.1.1 to 0.1.4.

I'm not intimately familiar with the kops way, so let me know if there's a better approach I should use or I've missed updating any generated files.

### CODE REVIEW

**Comment by ripta** [comment]
@chrislovecnm - I'm hoping you could take a look (or recommend someone else) if this seems sane to you. since you've done past reviews relating to the amazon-vpc CNI integration.

File: upup/pkg/fi/cloudup/template_functions.go [hunk 1]
```diff
@@ -106,6 +106,16 @@ func (tf *TemplateFunctions) AddTo(dest template.FuncMap) {
 		}
 		dest["FlannelBackendType"] = func() string { return flannelBackendType }
 	}
+
+	if tf.cluster.Spec.Networking != nil && tf.cluster.Spec.Networking.AmazonVPC != nil {
+		defaultPluginImageName := "602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni:0.1.4"
```
Comment (chrisz100) [inline upup/pkg/fi/cloudup/template_functions.go]: Setting customizable default values here and in nodeup doesn’t really look right to me. I’d personally move the default and customizabikity to the kops cluster spec networking portion.

Comment (ripta) [reply upup/pkg/fi/cloudup/template_functions.go]: @chrisz100 - Fair point! I'll take a look.

**Comment by ripta** [comment]
@chrisz100 - Thanks for taking a look at this PR. I pushed some changes:

1. Looking at the [calico](https://github.com/kubernetes/kops/blob/master/upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.7.yaml.template#L202-L216) and [canal](https://github.com/kubernetes/kops/blob/master/upup/models/cloudup/resources/addons/networking.projectcalico.org.canal/k8s-1.8.yaml.template#L142-L148) networking addons, looks like we set defaults in the templates, so I followed their lead.
2. I upgraded the plugin default to v1.0.0, which was released a few days ago.

Let me know what you think!

**Comment by chrisz100** [comment]
Looks better to me! Thanks :-) 
/cc @mikesplain remarks?

**Comment by mikesplain** [comment]
Also you should consider modifying the title of your PR to also mention the version bump, that way it shows up in the release notes better?

**Comment by mikesplain** [comment]
@ripta Thanks so much for the contribution. Since lots of people have started poking at the aws cni, I'm sure others will appreciate your work as well!

**Comment by ripta** [comment]
@mikesplain - Thanks for the comment! I fixed the version I had overlooked in source. I also moved the YAML template to follow the naming better (that is, to match the version of k8s compatible with the addon, rather than the version of the addon itself).

When testing and updating our manifest, I noticed one thing: [our version of the manifest](https://github.com/ripta/kops/blob/c3cc23b22aaab312d8d847725323b7a60ff729b5/upup/models/cloudup/resources/addons/networking.amazon-vpc-routed-eni/k8s-1.7.yaml.template#L56-L150) adds k8s-ec2-srcdst, while the [official version](https://github.com/aws/amazon-vpc-cni-k8s/blob/release-1.0/config/v1.0/aws-k8s-cni.yaml) doesn't (and never used) k8s-ec2-srcdst.

Do we _need_ k8s-ec2-srcdst? If we don't, I'd like to remove it. If we need it, I'd like to at least add a comment in the manifest. (Looks like @aledbf originally added it in #3997.)

**Comment by mikesplain** [comment]
@ripta Great thanks for the changes! 

That's a great point, I don't believe we need it, since all routing will be done by aws now... Have you tested removing it? We should certainly run a test before removing it. 

We can also just accept this as is and remove it in a separate PR to simplify things... thoughts?

**Comment by ripta** [comment]
@mikesplain - Yeah, I actually like that idea to address `srcdst` in a separate PR. I pushed a final commit that updates the manifest (path changes, env-var changes, new service account) to match the official manifest, *without* making changes to `srcdst`.

I just did an internal build of this latest change and a test deploy. The cluster came up healthy with the correct version of the image, so if the checks pass and changes look good to you, I think we're good with merging. :D

File: upup/models/cloudup/resources/addons/networking.amazon-vpc-routed-eni/k8s-1.7.yaml.template [hunk 2]
```diff
@@ -1,3 +1,44 @@
+# Vendored from https://github.com/aws/amazon-vpc-cni-k8s/blob/release-1.0/config/v1.0/aws-k8s-cni.yaml
+
+apiVersion: rbac.authorization.k8s.io/v1beta1
```
Comment (mikesplain) [inline upup/models/cloudup/resources/addons/networking.amazon-vpc-routed-eni/k8s-1.7.yaml.template]: Eventually we should probably create a second version of this manifest for 1.8+ with `rbac.authorization.k8s.io/v1`

**Comment by mikesplain** [comment]
Thanks for the contribution @ripta! 

@chrisz100 want to take another look at this since you originally reviewed this as well?

/lgtm

**Comment by chrisz100** [comment]
/lgtm 

Thank you for the contribution once again :-)

File: upup/models/cloudup/resources/addons/networking.amazon-vpc-routed-eni/k8s-1.7.yaml.template (hunk 2 above)
Comment (ripta) [reply upup/models/cloudup/resources/addons/networking.amazon-vpc-routed-eni/k8s-1.7.yaml.template]: Sounds good. I'll add it to my list.

File: pkg/apis/kops/networking.go [hunk 3]
```diff
@@ -124,6 +124,9 @@ type RomanaNetworkingSpec struct {
 
 // AmazonVPCNetworkingSpec declares that we want Amazon VPC CNI networking
 type AmazonVPCNetworkingSpec struct {
+	// The container image name to use, which by default is:
+	// 602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni:1.0.0
+	ImageName string `json:"imageName,omitempty"`
```
Comment (justinsb) [inline pkg/apis/kops/networking.go]: Nit: We probably shouldn't specify the default explicitly - rather we just allow overriding of the image name with this field.  Technically defaults are part of the API, and we're not guaranteeing this value as the default forever :-)

**Comment by justinsb** [comment]
I know there are lots of similar WIP PRs - I'm hoping this one is the newer version of those.  Apologies to anyone whose earlier work I misinterpreted or overlooked.

One nit that we want to avoid committing to a default value as part of the API contract, but we can come up with suitable wording later!

/approve
/lgtm

**Comment by mikesplain** [comment]
@xmubeta What version of kops are you using? This is only merged to master so in order to use it you'll have to build master and run it yourself.  I'd suggest waiting for the alpha to be cut, which should be soon!

**Comment by xmubeta** [comment]
Thank you, @mikesplain . I think you are right. I am using release version:
 Version 1.9.1 (git-ba77c9ca2)
That's why this feature is not included.

I can wait. Thank you again.

**Pull request merged**